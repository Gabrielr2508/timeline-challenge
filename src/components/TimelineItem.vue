<template>
  <div class="timeline-item"
    :class="icon"
  >
    <div class="content">
      <div class="header">
        <slot name="header"/>
      </div>
      <div class="body">
        <slot />
      </div>
    </div>
  </div>
</template>

<script>
export default {
  props: {
    status: {
      required: true,
      type: String,
      default: 'none',
    },
  },

  computed: {
    icon() {
      const availableIcons = ['success'];

      return availableIcons.includes(this.status)
        ? this.status
        : 'none';
    },
  },
};
</script>

<style lang="scss">
.timeline-item {
  padding: 10px 30px;
  position: relative;
  background-color: inherit;
  width: 50%;
  left: 33%;

  &.success {
    &::after {
      background-image: url('../assets/fonts/icons/check.svg');
      background-repeat: no-repeat;
      background-size: 1.5em;
    }
  }
  &::after { //circles
    content: "";
    position: absolute;
    width: 1.5em;
    height: 1.5em;
    right: -15px;
    top: 20px;
    border-radius: 50%;
    z-index: 1;
    left: -16px;
  }

  &::before { //arrows
    content: " ";
    height: 0;
    position: absolute;
    top: 22px;
    width: 0;
    z-index: 1;
    left: 20px;
    border: medium solid white;
    border-width: 10px 10px 10px 0;
    border-color: transparent white transparent transparent;
  }

  .content {
    // padding: 15px 20px;
    background-color: white;
    position: relative;
    border-radius: 6px;
    box-shadow: 0px 0px 10px 2px rgba(0,0,0,0.1);
    overflow: hidden;

    .header {
      padding: 10px 20px 10px 15px ;

      // height: 3em;
    }

    .body {
      background-color: #F8F8F8;
      padding: 10px 20px 10px 15px;
    }
  }
}

/* Media queries - Responsive timeline on screens less than 600px wide */
@media screen and (max-width: 600px) {
  .timeline-item  {
    width: 100%;
    padding-left: 60px;
    padding-right: 25px;
    left: 0%;

    &::before { //arrow
      left: 50px;
      border: medium solid white;
      border-width: 10px 10px 10px 0;
      border-color: transparent white transparent transparent;
    }

    &::after {
      left: 15px;
    }
  }
}
</style>
