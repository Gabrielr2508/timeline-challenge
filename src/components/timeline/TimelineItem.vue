<template>
  <div class="timeline-item-container">
      <div class="timeline-item"
        :class="icon"
      >
        <div class="content">
          <div class="header">
            <slot name="header"/>
          </div>
          <div class="body">
            <slot />
          </div>
        </div>
      </div>
  </div>
</template>

<script>
export default {
  props: {
    status: {
      required: true,
      type: String,
      default: 'none',
    },
  },

  computed: {
    icon() {
      const availableIcons = ['success'];

      return availableIcons.includes(this.status)
        ? this.status
        : 'none';
    },
  },
};
</script>

<style lang="scss" scoped>
.timeline-item-container {
  position: relative;
  max-width: 1000px;
  margin: 0 auto;

  &::after { //ruler
    content: '';
    position: absolute;
    width: 2px;
    background-color: #D8D8D8;
    top: 0;
    bottom: 0;
    left: 33%;
    margin-left: -3px;
  }
}

.timeline-item {
  padding: 10px 30px;
  position: relative;
  background-color: inherit;
  width: 50%;
  left: 33%;

  &.success {
    &::after {
      background-image: url('../../assets/fonts/icons/check.svg');
      background-repeat: no-repeat;
      background-size: 2em;
    }
  }
  &::after { //circles
    content: "";
    position: absolute;
    width: 2em;
    height: 2em;
    right: -15px;
    top: 20px;
    border-radius: 50%;
    z-index: 1;
    left: -16px;
  }

  &::before { //arrows
    content: " ";
    height: 0;
    position: absolute;
    top: 24px;
    width: 0;
    z-index: 1;
    left: 20px;
    border: medium solid white;
    border-width: 10px 10px 10px 0;
    border-color: transparent white transparent transparent;
  }

  .content {
    // padding: 15px 20px;
    background-color: white;
    position: relative;
    border-radius: 6px;
    box-shadow: 0px 0px 10px 2px rgba(0,0,0,0.1);
    overflow: hidden;

    .header {
      padding: 10px 20px 10px 20px ;
    }

    .body {
      background-color: #F8F8F8;
      padding: 10px 20px 10px 20px;
    }
  }
}

/* Media queries - Responsive timeline on screens less than 600px wide */
@media screen and (max-width: 600px) {
  .timeline-item  {
    width: 100%;
    padding-left: 60px;
    padding-right: 25px;
    left: 0%;

    &::before { //arrow
      left: 50px;
    }

    &::after {
      left: 15px;
    }
  }

  .timeline-item-container::after {
    left: 31px;
  }

}
@media screen and (min-width: 600px) and (max-width: 768px) {
  .timeline-item  {
    width: 80%;
    padding-left: 60px;
    padding-right: 25px;
    left: 10%;

    &::before { //arrow
      left: 50px;
    }

    &::after { //circles
      left: 15px;
    }
  }

  .timeline-item-container::after { //ruler
    left: 10%;
    margin-left: 28px;

  }
}

@media screen and (min-width: 768px) and (max-width: 992px) {
  .timeline-item  {
    width: 70%;
    padding-left: 60px;
    padding-right: 25px;
    left: 15%;

    &::before { //arrow
      left: 50px;
    }

    &::after { //circles
      left: 15px;
    }
  }

  .timeline-item-container::after { //ruler
    left: 15%;
    margin-left: 28px;

  }
}
</style>
