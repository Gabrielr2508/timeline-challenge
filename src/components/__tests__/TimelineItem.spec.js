import { shallowMount } from '@vue/test-utils';

import TimelineItem from '../TimelineItem.vue';

describe('TimelineItem.vue', () => {
  let wrapper;

  beforeEach(() => {
    const props = {
      status: 'success',
    };

    wrapper = shallowMount(TimelineItem, {
      propsData: { ...props },
      slots: {
        header: '<h5>Dummy header</h5>',
        default: '<div class="fake-info">Fake text</div>',
      },
    });
  });

  test('header slot is inserted in a .timeline-item .header element', () => {
    const list = wrapper.find('.timeline-item .header');

    expect(list.findAll('h5').length).toBe(1);
  });

  test('default slot is inserted in a .timeline-item .content element', () => {
    const list = wrapper.find('.timeline-item .content');

    expect(list.findAll('.fake-info').length).toBe(1);
  });

  test('put css icon class according to the status prop', () => {
    let list = wrapper.findAll('.timeline-item.success');

    expect(list.length).toBe(1);

    wrapper.setProps({ status: 'anything' });

    list = wrapper.findAll('.timeline-item.none');
    expect(list.length).toBe(1);
  });

  test('renders TimelineItem correctly', () => {
    expect(wrapper.element).toMatchSnapshot();
  });
});
