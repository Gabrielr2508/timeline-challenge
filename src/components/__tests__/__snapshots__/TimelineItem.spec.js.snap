// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimelineItem.vue renders TimelineItem correctly 1`] = `
<div
  class="timeline-item success"
>
  <div
    class="header"
  >
    <h5>
      Dummy header
    </h5>
  </div>
   
  <div
    class="content"
  >
    <div
      class="fake-info"
    >
      Fake text
    </div>
  </div>
</div>
`;
